statsmodels/base/distributed_estimation.py

```diff
@@ -0,0 +1,400 @@
+
+    return beta_tilde
+#    return beta_tilde, beta_mn, theta_hat, grad_mn, gamma_hat_l, theta_hat.dot(grad_mn)
+
+
+def distributed_estimation(data_generator, partitions,
+                           model_class=None, init_kwds=None, fit_kwds=None,
+                           estimation_method=None, estimation_kwds=None,
+                           join_method=None, join_kwds=None):
+    """This functions handles a general approach to distributed estimation,
+    the user is expected to provide generators for the data as well as
+    a model and methods for performing the estimation and recombining
+    the results
+
+    Parameters
+    ---------
+    data_generator : generator
+        A generator that produces a sequence of tuples where the first
+        element in the tuple corresponds to an endog array and the
+        element corresponds to an exog array.
+    partitions : scalar
+        The number of partitions that the data will be split into.
+    model_class : statsmodels model class
+        The model class which will be used for estimation. If None
+        this defaults to OLS.
+    init_kwds : dict-like or None
+        Keywords needed for initializing the model, in addition to
+        endog and exog.
+    fit_kwds : dict-like or None
+        Keywords needed for the model fitting.
+    estimation_method : function or None
+        The function that performs the estimation for each partition.
+        If None this defaults to _est_regularized_distributed.
+    estimation_kwds : dict-like or None
+        Keywords to be passed to estimation_method.
+    join_method : function or None
+        The function used to recombine the results from each partition.
+        If None this defaults to _join_debiased.
+
+    Returns
+    -------
+
+    join_method result.  For the default _join_debiased, it returns a
+    p length array.
+    """
+
+    init_kwds = {} if init_kwds is None else init_kwds
+    estimation_kwds = {} if estimation_kwds is None else estimation_kwds
+    join_kwds = {} if join_kwds is None else join_kwds
+
+    # set defaults
+    if model_class is None:
+        model_class = OLS
+
+    if estimation_method is None:
+        estimation_method = _est_regularized_distributed
+
+    if join_method is None:
+        join_method = _join_debiased
+
+    model_results_l = []
+
+    # index for machine
+    mnum = 0
+
+    # TODO given that we already have an example where generators should
+    # produce more than just exog and endog (partition for variables)
+    # this should probably be handled differently
+    for endog, exog in data_generator:
+
+        model = model_class(endog, exog, **init_kwds)
+
+        # TODO possibly fit_kwds should be handled within
+        # estimation_kwds to make more general?
+        results = estimation_method(model, mnum, partitions, fit_kwds,
+                                    **estimation_kwds)
+        model_results_l.append(results)
+
+        mnum += 1
```

Comment:
I like that it's all separated out in functions, that makes the pattern here easier to understand.
However, AFAICS this is not yet distributed computation, it's out of core where the data needs to come to the central function.
My guess is that this loop could work in parallel using joblib. We have it as optional dependency, but I haven't looked at it in a few years.